File: python/day15/day15.py
import pathlib
import sys
from collections import OrderedDict, defaultdict


def parse(parsedata):
    assert r_hash("HASH") == 52
    return parsedata.split(",")


def r_hash(line):
    curr = 0
    for el in line:
        curr = ((curr + ord(el)) * 17) % 256
    return curr


def part1(data):
    return sum(r_hash(line) for line in data)


def part2(data):
    hashmap = defaultdict(OrderedDict)

    for label in data:
        if "=" in label:
            _key, focal_length = label.split("=")
            hashmap[r_hash(_key)][_key] = int(focal_length)
        elif "-" in label:
            _key = label.split("-")[0]
            hashmap[r_hash(_key)].pop(_key, None)

    result = 0
    for _key in hashmap:
        for i, focal_length in enumerate(hashmap[_key].values(), start=1):
            print(_key, hashmap[_key].items())
            print(f"curr result {result} + new value {(1 + _key) * i * focal_length}")
            result += (1 + _key) * i * focal_length

    print(hashmap)
    return result


def solve(puzzle_data):
    data = parse(puzzle_data)
    solution1 = part1(data)
    solution2 = part2(data)
    return solution1, solution2


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"{path}:")
        puzzle_input = pathlib.Path(path).read_text().strip()
        solutions = solve(puzzle_input)
        print("\n".join(str(solution) for solution in solutions))

File: 2023/csharp/day15/day15.cs
using System.Collections.Specialized;

class SolveDay15(String path)
{
    private readonly string[] input = File.ReadAllText(path).Trim().Split(',');

    private static int Hash(string line)
    {
        int curr = 0;
        foreach (char c in line)
        { curr = (curr + ((int)c)) * 17 % 256; }
        return curr;
    }


    public int Part1()
    {
        return this.input.Select(Hash).Sum();
    }

    public int Part2()
    {
    Dictionary<int, OrderedDictionary> hashmap = [];

        foreach (string label in this.input)
        {
            if (label.Contains('='))
            {
                string[] parts = label.Split("=");
                string _key = parts[0];
                int focal_length = int.Parse(parts[1]);
                int hash = Hash(_key);
                if (!hashmap.ContainsKey(hash))
                {
                    hashmap[hash] = [];
                }
                hashmap[hash][_key] = focal_length;
            }
            else if (label.Contains('-'))
            {
                string _key = label.Split("-")[0];
                int hash = Hash(_key);
                if (hashmap.ContainsKey(hash))
                {
                    hashmap[hash].Remove(_key);
                }
            }
        }

        int result = 0;
        foreach (int key in hashmap.Keys)
        {
            int i = 1;
            foreach (int focal_length in hashmap[key].Values)
            {
                result += (1 + key) * i * focal_length;
                i++;
            }
        }

        return result;
    }

}